\chapter{Biological background}
In this chapter we will present some basic biological terms, which we will need later in this work.
At first we will describe the structure of DNA, its function in living organisms
and methods of representation in bioinformatics and computational biology.
We will also briefly describe how we can get information about DNA from living organisms and we will mention central dogma of molecular biology, which is needed in order to understand relationships between biological macromolecules.

\section{What is DNA}
Deoxyribonucleic acid (DNA) is a natural polymer, a long chain of nucleotides. There are 4 types of nucleotides
naturally occurring in DNA - adenine, cytosine, guanine and thymine. These bases are connected through phosphodiester bonds,
connection between the 3' carbon atom of one deoxyribose and the 5' carbon atom of the second deoxyribose. \\
The structure of DNA consists of 2 complementary strands coiled around each other in the form of double helix.
In this double helix adenine is paired with thymine and cytosine with guanine through hydrogen bonds.
In bioinformatics and computational biology we usually represent a base by its first letter - A for adenine, C for cytosine, G for guanine and T for thymine.
We usually store DNA sequences in simple plain text files. Probably the most widely used format is FASTA format.
It can consist of one or multiple records, where each record starts with one line for record id and comment, which is followed by possibly multiple lines of sequence writen in direction from 5' to 3'.

\section{DNA sequencing}
DNA sequencing is a process of retrieving data from biological samples.
Data of DNA are usually retrieved in a small fragments on account of limitations of technology.
Currently there are 3 main technologies of DNA sequencing - Sanger sequencing, NGS sequencing and Nanopore sequencing.
They differ in key parameters, such as length of read, error rate, throughput and price per base.
We show a table of comparison between these techniques.

\begin{center}
  \begin{tabular}{ | l | c | c | c | c | }
    \hline
    method & length of read & error rate & throughput per day & price per MB \\
    \hline
    Sanger & 1000bp & <2\% & 3MB & 4000\$ \\
    NGS & 2$\times$150bp & <1\% & 30GB & <1\$ \\
    Nanopore & 100kbp & 15\% & 1GB & 1\$ \\
    \hline
  \end{tabular}
\end{center}

\section{Central dogma of molecular biology}
We already showed some basic concepts about what DNA is and how we can work with it.
In this section we would like to present how DNA is interpreted in living organisms.
Some parts of DNA are transcribed into RNA molecules.
RNA is, similarly as DNA, a long chain of nucleotides, but there are some important differences.
The differences in these 2 molecules are, that RNA is mostly found in nature as single stranded molecule.
Nucleotides in RNA also contain ribose instead of deoxyribose and nucleotide thymine is altered for uracil.
RNA sometimes also exhibits enzymatic activity, but mostly it is translated into protein.
Proteins are long chains of amino acids and can be presented in computer in similar way as DNA or RNA, althought there is 20 proteinogenic amino acids.
We provide a table of codes for different amino acids.
Proteins have highly variable functions and therefore they will be of high importance in our analysis.

\medskip
(Amino acids table)
\medskip





